### PR TITLE
Add progress goal checker ( issue #6033 )

### LINKED ISSUE #6033: New FallbackToleranceGoalChecker?
#### Feature description

What do we think about a new goal checker that has two tolerance tiers — a tight "desired" tolerance and a looser "minimum" tolerance — so that:

If the robot reaches within the desired tolerance, the goal is marked complete immediately
If the robot is within the minimum tolerance but hasn't made meaningful progress toward the goal for some configurable duration, the goal is marked complete — accepting "good enough" rather than fighting forever to close that last few centimeters.

I got the idea because with controllers like MPPI, they can most of the time reach a goal precisely but sometimes start oscillating close to it - This new goal checker would allow to reach the goal precisely most of the time but also succeed in the rare cases where we are "close enough" and can't make more progress

#### Implementation considerations

Should this be a GoalChecker plugin or a more elaborate BT Implementation?

P.S: The name `FallbackToleranceGoalChecker` can be improved

### LINKED ISSUE #6099: Can DWB be tuned to penetrate deeper into the XY goal region before final rotation, without changing xy_goal_tolerance?
Hi all,

I am doing a sim-to-real analysis of MPPI vs DWB on a skid-steer robot in a narrow warehouse task, and I am trying to understand whether DWB can be made to drive deeper into the XY goal region before final rotation, without tightening DWB’s own internal near-goal tolerance.

Context
- Platform: skid-steer / differential-drive robot
- Stack: Nav2 Jazzy
- Controllers compared: DWB vs MPPI
- Same robot, same task, same localization pipeline, same goals, same goal checker
- Goal checker: `SimpleGoalChecker(xy_goal_tolerance=0.10, yaw_goal_tolerance=0.05, stateful=false)`
- I am explicitly comparing native controller behaviour, so I am not using Rotation Shim or controller switching

Observed behaviour in the controller-visible map frame
- DWB sim: 29/30 success
- DWB real: 1/15 success
- MPPI sim: 29/30 success
- MPPI real: 15/15 success

Near-goal metrics:
- DWB median `min_xy_err`: 0.094 m in sim, 0.105 m in real
- MPPI median `min_xy_err`: 0.023 m in sim, 0.030 m in real

What I consistently see is:
- DWB tends to operate right at the XY tolerance boundary
- Near goal, DWB often becomes rotate-dominant while still having very little XY margin
- MPPI tends to drive much deeper into the goal region before final orientation correction

This makes DWB much more sensitive to small real-world effects such as skid-steer motion behaviour, wheel separation / ICR mismatch, low-speed rotation effects, and localization imperfections.

What I can already do
I was able to make DWB converge much deeper by reducing DWB’s internal `xy_goal_tolerance` from `0.10` to `0.05`, while keeping the external goal checker at `0.10`.

In a runtime single-goal probe, that gave:
- final_xy_error: `0.057 m`
- min_xy_error_est: `0.053 m`
- success: `True`

So tightening the controller’s internal near-goal threshold clearly changes the behaviour in the direction I want.

My actual question
I do NOT mainly want to know whether DWB can be made deeper by changing its internal `xy_goal_tolerance`, because I have already confirmed that.

What I want to understand is:

Can DWB be made to converge deeper before final rotation while keeping its internal `xy_goal_tolerance` unchanged?

In other words:
- is deeper XY penetration fundamentally tied to the internal threshold at which DWB switches into terminal behaviour
- or are there supported ways to get similar behaviour from the algorithm itself, without moving that threshold?

More concretely:
1. Is DWB fundamentally expected to stop around its internal `xy_goal_tolerance` and then rotate?
2. If I keep that internal threshold unchanged, are there tuning paths that can still make DWB translate deeper before final rotation?
3. If yes, which parameters matter most?
   - `RotateToGoal`
   - `slowing_factor`
   - `GoalDist`, `GoalAlign`, `PathAlign`
   - sampling / horizon settings
   - any other near-goal mechanism
4. Or is this effectively a structural limitation of DWB, where deeper terminal convergence mostly requires changing the internal threshold itself?

What I already tried
- Changing critic weights
- Adjusting `RotateToGoal` related settings
- Trying to bias the controller toward stronger translational convergence near goal

So far, the only thing that clearly changes the behaviour in the desired way is reducing DWB’s internal `xy_goal_tolerance`.

That is why my main question is whether I am missing an intended algorithmic tuning path, or whether this is simply not how DWB is designed to behave.

If useful, I can also share:
- a minimal config
- the runtime single-goal probe output
- short near-goal trajectory plots comparing DWB and MPPI

### PR DESCRIPTION
**Pull request by DavidG-Develop**
Please advise if you have better name for the plugin due to possible confusion with progress checkers.

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #6033  |
| Primary OS tested on | Ubuntu, dockerized |
| Robotic platform tested on | TB4 Sim |
| Does this PR contain AI generated software? | Yes, AI assisted |
| Was this PR description generated by AI software? | No |

---

## Description of contribution in a few bullet points

Added goal checker based on #6033. Goal checker with a tight and a loose tolerance bound. Tight tolerance bound instantly returns goal reached while the loose tolerance bound waits if the robot does not make progress for X cycles while within bound than succeed.

Based on simple goal checker improved by using the second bound with the parameter for number of cycles when to succeed.

## Description of documentation updates required from your changes

Should be added to goal checker list.

## Description of how this change was tested

Wrote tests for both bounds as well as robot improving towards the goal and not and tested in sim with goals close to obstacles.

---

## Future work that may be required in bullet points

Name could be confusing due to progress checkers but also makes sense for this plugin :/

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

### CODE REVIEW

File: nav2_controller/include/nav2_controller/plugins/progress_goal_checker.hpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+protected:
+  nav2::LifecycleNode::WeakPtr node_;
+  rclcpp::Logger logger_{rclcpp::get_logger("progress_goal_checker")};
+
+  // Tight (desired) tolerance
+  double xy_goal_tolerance_;
+  double xy_goal_tolerance_sq_;
+  // Loose (fallback) tolerance
+  double max_xy_goal_tolerance_;
+  double max_xy_goal_tolerance_sq_;
```
Comment (SteveMacenski) [inline nav2_controller/include/nav2_controller/plugins/progress_goal_checker.hpp]: I think renaming these to be more clear would be good: coarse and fine goal tolerances. Those are common words used for this kind of thing

File: nav2_controller/plugins/progress_goal_checker.cpp [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+{
+  plugin_name_ = plugin_name;
+  node_ = parent;
+  auto node = node_.lock();
+  logger_ = node->get_logger();
+
+  xy_goal_tolerance_ = node->declare_or_get_parameter(
+    plugin_name + ".xy_goal_tolerance", 0.10);
+  max_xy_goal_tolerance_ = node->declare_or_get_parameter(
+    plugin_name + ".max_xy_goal_tolerance", 0.25);
```
Comment (SteveMacenski) [inline nav2_controller/plugins/progress_goal_checker.cpp]: Ditto here with coarse and fine

File: nav2_controller/plugins/progress_goal_checker.cpp [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+      RCLCPP_INFO(
+        logger_,
+        "ProgressGoalChecker: accepting goal at tight tolerance "
+        "(current: %.3f m, tol: %.3f m)",
+        std::sqrt(dist_sq), xy_goal_tolerance_);
+      // Fall through to yaw check
+    } else if (dist_sq <= max_xy_goal_tolerance_sq_) {
+      // Tier 2: Within the loose tolerance zone — track convergence
+      if (!in_tolerance_zone_) {
+        // Just entered the zone: initialize tracking
```
Comment (SteveMacenski) [inline nav2_controller/plugins/progress_goal_checker.cpp]: Move comment here and in the other conditionals before the `if...` line (L151)

File: nav2_controller/plugins/progress_goal_checker.cpp [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        no_improvement_count_ = 0;
+        return false;
+      }
+
+      if (dist_sq < best_distance_sq_) {
+        // Robot is still converging: update best, reset counter
+        best_distance_sq_ = dist_sq;
+        no_improvement_count_ = 0;
+        return false;
+      }
```
Comment (SteveMacenski) [inline nav2_controller/plugins/progress_goal_checker.cpp:189]: My major concern here is 2 fold:
* What if its barely inching forward because its trying to get that last little bit but its not really productive refinement?
* If there's localization error because someone's using something noisy like VIO this could be triggered many times just from jitter

We also have access to information about the velocity of the vehicle in this API. Seems like it would be wise to use this to see if the robot is just simply stopped, stalled, or otherwise below the threshold of doing "useful" things. That would help resolve largely both points I think. Check out the Stopped Progress Checker for reference.

File: nav2_controller/plugins/test/goal_checker.cpp
Comment (SteveMacenski) [inline nav2_controller/plugins/test/goal_checker.cpp:1]: I would recommend reviewing these tests and seeing if they can't be made more ... human looking and concise. Not saying these are wrong, but I would ask you to validate that this is something you'd put your name on if these are tests written by AI

**Review by SteveMacenski — COMMENTED** [review]
On names: I agree this should be adjusted
* CoarseToFineGoalChecker
* AdaptiveToleranceGoalChecker

The usual (probably after we land on a name): 
* Update the migration guide to mention to folks this exists
* Update the configuration guide with a new entry for this plugin
* Update the plugins page to list this a progress checker option

File: nav2_controller/plugins/progress_goal_checker.cpp (hunk 4 above)
Comment (DavidG-Develop) [reply nav2_controller/plugins/progress_goal_checker.cpp:189]: Yep, makes sense. With the assumption that you meant Stopped Goal Checker and not progress checker and checking its code I see it is a plain if speed < threshold, would you than recommend: 

- The same approach where we than instantly return goal reached within the fine bound and within the coarse bound we check if speed < threshold over the set amount of cycles? This would keep the code simple just by switching the geometry check to a velocity check (could be impacted by a large controller level oscillations but that is than a controller tuning problem and not a goal checker problem)
- Or some additional check if the velocity is actually "useful"? Something like a velocity projection checking if the velocity is "pointing" towards the goal pose next to the check if the speed < threshold where we can than keep the threshold lower to detect stalling

File: nav2_controller/plugins/test/goal_checker.cpp
Comment (DavidG-Develop) [reply nav2_controller/plugins/test/goal_checker.cpp:1]: Yep, Ill take a look... They are AI generated, I did go through all of them and they do cover all cases but you have 100% right that they also look robotic... When I redo the goal checking with velocity I will have to rewrite them as well and I will see if there is a better way to write them...

**Comment by DavidG-Develop** [comment]
My vote would go towards AdaptiveToleranceGoalChecker! As soon as we agree on the name (and the implementation specifics) I will rename it and look towards implementing the doc changes

**Comment by SteveMacenski** [comment]
`AdaptiveToleranceGoalChecker` is it!

File: nav2_controller/plugins/progress_goal_checker.cpp (hunk 4 above)
Comment (SteveMacenski) [reply nav2_controller/plugins/progress_goal_checker.cpp:189]: Indeed what I meant :wink: 

I think if we're in the coarse bounds and stopped (`< threshold`), consider it done with the same `no_improvement_count_` cycle counter (might want to be renamed to `completed_state_counter_` or something to be more generic).  

The problem is some controllers if you let it keep going like MPPI will do small refinements and jitter to try to continue towards the goal objective function, even though its so small its really not doing much useful work. Similarly, if the controller is publishing something very small which the robot is not actually able to accomplish due to stall torques and/or going up-hill at a low speed, we also want to terminate if we're in the coarse tolerance. 

We want to terminate if we're not actually doing much of anything like < 0.1m/s for a long time. The key here is that we would want this to be sufficiently low that we don't just call the goal accomplished if we're crawling towards the goal with meaningful progress -- but also stop it proactively when we're not actually doing much. 

Setting a low "stall" velocity limit seems like a good way. But may also consider doing something like you suggest with analyzing the recent window of velocities and positions to see if making useful progress. That may be overkill though.

**Comment by SteveMacenski** [comment]
P.S. Not necessarily asking you to make this change yet, but something related to this is this function: https://github.com/ros-navigation/navigation2/blob/main/nav2_route/src/route_tracker.cpp#L54

In this, I check if we pass a tangental line at the goal pose orthogonal to the goal orientation. We call the node achieved not only when we're within tolerance, but when we actually pass that tangential line marking the goal-aligned pose has been passed (with whatever latitudinal error). 

We may be able to use that kind of logic here as another "we're done" condition. If we are in the coarse window **and** we have just passed the goal pose "finish line", we should stop. 

What do you think? 

I'm not 100% sure at a lobbing-out-an-idea view, but that may replace the need for `best_distance_sq_` entirely? Maybe good to do both, actually, for the case that it stops before it actually crosses that "finish line"

**Comment by DavidG-Develop** [comment]
So I currently just implemented it as simple as possible by taking how the Stopped goal checker works and simply switched the current geometry approach for a velocity approach. This could already be further tested, I think @jkeo180 from the issue showed interest in helping maybe she can test it and maybe also give further ideas into possible improvements.

I would now take a look in the next couple of days ( as I do have some other stuff to do like my job :) ) which kinds of checks can we do for "usefullsness" and take a look at the route tracker implementation of node achieved to see how could we best implement it here. I will try to tackle it over the weekend or latest start of next week. Until than we can also further discuss which solution might fit best here.  Of course as always all theoretical and implementation guidelines are very welcome!

File: nav2_controller/plugins/test/goal_checker.cpp
Comment (SteveMacenski) [reply nav2_controller/plugins/test/goal_checker.cpp:1]: did you do this yet?

File: nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    if (dist_sq <= fine_xy_goal_tolerance_sq_) {
+      if (stateful_) {
+        check_xy_ = false;
+      }
+      RCLCPP_INFO(
+        logger_,
+        "AdaptiveToleranceGoalChecker: accepting goal at fine tolerance "
+        "(current: %.3f m, tol: %.3f m)",
+        std::sqrt(dist_sq), fine_xy_goal_tolerance_);
+      // Fall through to yaw check
```
Comment (SteveMacenski) [inline nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp]: If stateful = false and rotating to heading, this could print at control_freq rate. I think this needs to print at the very end with an if/else for if using the fine or coarse after the yaw check succeeds

File: nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+      // Check if the robot is effectively stopped
+      const bool robot_stopped =
+        std::hypot(velocity.linear.x, velocity.linear.y) <= trans_stopped_velocity_ &&
+        std::fabs(velocity.angular.z) <= rot_stopped_velocity_;
+
+      if (robot_stopped) {
+        stagnation_count_++;
+      } else {
+        stagnation_count_ = 0;
+      }
```
Comment (SteveMacenski) [inline nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp]: What happened to the distance check before? This is not either-or but both I think. If we're not making progress towards the goal (best distance isn't improving), we also need to set the counter, right? Seems like a good hedge to handle both cases.

File: nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp (hunk 5 above)
Comment (DavidG-Develop) [reply nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp]: Yep correct, will move down to the end so it only logs once always.

File: nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp (hunk 6 above)
Comment (DavidG-Develop) [reply nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp]: Oh okay my bad, I thought this was the wanted functionality when you mentioned the stopped checker. Now i returned what we had and combined velocity + distance checking where if its stopped or no imrovement in distance we increase the counter

**Comment by DavidG-Develop** [comment]
I did also redo the tests you can take a look if it looks better.

File: nav2_controller/include/nav2_controller/plugins/adaptive_tolerance_goal_checker.hpp [hunk 7]
```diff
@@ -116,6 +116,8 @@ class AdaptiveToleranceGoalChecker : public nav2_core::GoalChecker
   bool check_xy_;
   bool in_tolerance_zone_;
   int stagnation_count_;
+  double best_distance_sq_;
+  bool accepted_at_fine_;
```
Comment (SteveMacenski) [inline nav2_controller/include/nav2_controller/plugins/adaptive_tolerance_goal_checker.hpp]: I don't think this needs to be a class member :-)

File: nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+        in_tolerance_zone_ = true;
+        stagnation_count_ = 0;
+        best_distance_sq_ = dist_sq;
+        return false;
+      }
+
+      // Check if the robot is stopped or not making progress toward goal
+      const bool robot_stopped =
+        std::hypot(velocity.linear.x, velocity.linear.y) <= trans_stopped_velocity_ &&
+        std::fabs(velocity.angular.z) <= rot_stopped_velocity_;
```
Comment (SteveMacenski) [inline nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp:161]: What happened to the velocity check if we're stopped to also trigger the stagnation count? We want **both** I think. If we're stopped for a long time or not making progress; stop. If we're not making `distance_improved` also stop. These should be decoupled. 

I feel like we're playing wack-a-mole here 😆

File: nav2_controller/include/nav2_controller/plugins/adaptive_tolerance_goal_checker.hpp (hunk 7 above)
Comment (DavidG-Develop) [reply nav2_controller/include/nav2_controller/plugins/adaptive_tolerance_goal_checker.hpp]: True, thats what happens when I say I wont do it and than I end up doing it quickly and not thinking it through :)

File: nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp (hunk 8 above)
Comment (DavidG-Develop) [reply nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp:161]: Wack-a-mole is good hand eye coordination practice :wink:

**Comment by DavidG-Develop** [comment]
I thought about the route tracker implementation you mentioned... Just to make sure we are on the same page implementation should run still side by side to the current stagnation implementation but with the extra nice to have of triggering instantly on crossing the "goal line"?

Something like 
finish_line_concept

Now my questions would be: 

1. Does this actually make sense as it is possble the robot will trigger this before trying to actually correct in case of overshooting
2. How to tackle if the robot is already approaching from the dot > 0 side due to us using navfn? Do we just track last known dot and if we go from < 0 -> > 0 than trigger but if already come from the side with > 0 just fall through to the stagnation checks?

Also btw my commits might be slow this week as I have quite a busy week. :frowning_face:

**Comment by SteveMacenski** [comment]
> Just to make sure we are on the same page implementation should run still side by side to the current stagnation implementation but with the extra nice to have of triggering instantly on crossing the "goal line"?

Yeah, precisely. The stagnation though should consider the change in distance and the velocity though (not as linked, but as 2 things that can both increment the counter).

>   Does this actually make sense as it is possble the robot will trigger this before trying to actually correct in case of overshooting

I think so? I'm not sure that some small correction will look different from a small velocity for the stagnation checks to get it. I think lets not try to over optimize here, we have enough going on already. I think if stopped / uselessly small velocity / jitter within coarse --> accept. If within coarse but moving away now with a counter --> accept. If in coarse and passed the "finish line" --> accept. If in fine at all --> accept. 

Folks can add more in the future form there. I think that handles the realistic coarse-window optimizations that I can think of (stopped already / not making progress / instantly passed the line to bypass the stagnation counter so we don't continue to drive away once passed the tangent line). 

> How to tackle if the robot is already approaching from the dot > 0 side due to us using navfn? Do we just track last known dot and if we go from < 0 -> > 0 than trigger but if already come from the side with > 0 just fall through to the stagnation checks?

Maybe use the path approach orientation (angle from pose `N` and `N-1` -- though probably not `N-1` due to NavFn jitter. Maybe 3-5?) rather than the orientation of the goal pose itself. 

Good catch.

**Comment by DavidG-Develop** [comment]
> Yeah, precisely. The stagnation though should consider the change in distance and the velocity though (not as linked, but as 2 things that can both increment the counter).
> 

Yep that was the last implementation, please check if it is fine at its current state (2 seperate counters 1 for stopped one for distance stagnation).

>  I think if stopped / uselessly small velocity / jitter within coarse --> accept. If within coarse but moving away now with a counter --> accept. If in coarse and passed the "finish line" --> accept. If in fine at all --> accept.
> 
> Maybe use the path approach orientation (angle from pose `N` and `N-1` -- though probably not `N-1` due to NavFn jitter. Maybe 3-5?) rather than the orientation of the goal pose itself.
> 

Okay I will take a look into implementing than this extra 3rd success trigger. I would do the N-x x as parameter just so we can also adapt to other planners when needed. I hope to get the commit out in the next couple of days.

Also please take a look at the current set of test. Once I implement what we discussed I will just add tests for the 3rd accept trigger in the same fashion.

The release test keeps failing due to it looking for the old file `Could not find '/opt/overlay_ws/src/navigation2/nav2_controller/plugins/progress_goal_checker.cpp' to scan for exclusion markers...` should I do something about that?

**Comment by DavidG-Develop** [comment]
Okay so I tried to make it planner agnostic so that I take the first position within the coarse zone and use that to compute the finish line. If I am not mistaken this should handle what we want nicely and work with "all" planners. Maybe I skipped some edge case that this does not cover so please advise.

Here is than also a sketch of all 4 success paths possible (might be also good to add into the documentation later)

adaptive_tolerance_viz

File: nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+    // Tier 2: Within the coarse tolerance zone — check velocity stagnation
+    } else if (dist_sq <= coarse_xy_goal_tolerance_sq_) {
+      // Just entered the zone: initialize tracking
+      if (!in_tolerance_zone_) {
+        in_tolerance_zone_ = true;
+        stopped_stagnation_count_ = 0;
+        distance_stagnation_count_ = 0;
+        best_distance_sq_ = dist_sq;
+        approach_dx_ = -dx;
+        approach_dy_ = -dy;
```
Comment (SteveMacenski) [inline nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp:169]: Why negative? I'm not sure I track that

File: nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+        return false;
+      }
+
+      // Finish line: robot crossed from approaching (dot<0) to passed (dot>=0)
```
Comment (SteveMacenski) [inline nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp]:
```suggestion
      // Finish line: robot crossed from approaching (dot<0) to passed (dot>=0) the
      // virtual line orthogonal to the approach pose and the goal pose, representing 
      // that the robot has passed the goal longitudinally while approaching.
```

File: nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+      const double finish_dot = dx * approach_dx_ + dy * approach_dy_;
+      const bool crossed_finish_line = finish_dot >= 0.0;
```
Comment (SteveMacenski) [inline nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp]:
```suggestion
      const bool crossed_finish_line = dx * approach_dx_ + dy * approach_dy_ >= 0.0;
```

File: nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+        std::hypot(velocity.linear.x, velocity.linear.y) <= trans_stopped_velocity_ &&
+        std::fabs(velocity.angular.z) <= rot_stopped_velocity_;
+
+      // Check if best distance has been improved
+      const bool distance_improved = dist_sq < best_distance_sq_;
+
+      if (distance_improved) {
+        best_distance_sq_ = dist_sq;
+      }
+
```
Comment (SteveMacenski) [inline nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp]:
```suggestion
      if (distance_improved) {
        best_distance_sq_ = dist_sq;
        distance_stagnation_count_ = 0;
      } else {
        distance_stagnation_count_++;
      }
```

Then remove the checks in L186 below also looking at the distance improved? If so, then this can also be compacted further as `if (dist_sq < best_distance_sq_)` and remove the variable `distance_improved`

File: nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+      if (robot_stopped) {
```
Comment (SteveMacenski) [inline nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp]: Just do the velocity checks in the `if(...)` statement :-) Keeps it all compact

**Comment by SteveMacenski** [comment]
> Okay so I tried to make it planner agnostic so that I take the first position within the coarse zone and use that to compute the finish line. If I am not mistaken this should handle what we want nicely and work with "all" planners. Maybe I skipped some edge case that this does not cover so please advise.

Clever, I like that

> Here is than also a sketch of all 4 success paths possible (might be also good to add into the documentation later)

Great thanks! That'll be good for the configuration guide docs on this :-)

File: nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp (hunk 9 above)
Comment (DavidG-Develop) [reply nav2_controller/plugins/adaptive_tolerance_goal_checker.cpp:169]: Due to dx and dy being vector goal -> robot than this would be vector robot to goal which allows for `finish_dot = dx * approach_dx_ + dy * approach_dy_` can allow me to check `crossed_finish_line = finish_dot >= 0.0`. I could switch either this to be `approach_dx_ = dx, approach_dy_ = dy`, than either add a negation for the dot calculation or `crossed_finish_line = finish_dot <= 0.0` but I wanted to set that passed is when `>=0.0`. Please correct if my math is wrong :laughing:

**Comment by tonynajjar** [comment]
Nice thanks for working on that, it looks more complex than what I anticipated when I wrote the issue, but I trust that it's for good reason. I won't be able to review it in detail but will give it a test sometime next week.

What about adding tests? I think the complexity warrants some.

@SteveMacenski what do you think about this becoming the nav2 default?

**Comment by DavidG-Develop** [comment]
@tonynajjar I did also add test here https://github.com/ros-navigation/navigation2/pull/6052/changes#diff-4a74fd3e34ade379d7ab0aea85ef84fc620ff36d10d8c1c3389b68e5877a9747R471-R872

Steve did ask about redoing them at the beginning, which I did, after which we did not comment on if they look good or not. The tests are made to cover all 3 success cases (it is very possible that I forgot some edge cases so please check)

**Comment by tonynajjar** [comment]
apologies, I was blind :sweat_smile:

**Comment by SteveMacenski** [comment]
> @SteveMacenski what do you think about this become the nav2 default?

After you give it a whirl and are happy with it, I'm OK with that! I tentatiely approve pending your review

**Comment by jkeo180** [comment]
Hey @DavidG-Develop, I'm checking the logic now. The 'finish line' concept is exactly what I was hoping for to prevent the robot from oscillating at the edge of the goal.

**Comment by tonynajjar** [comment]
What I did to test:

- Set the `fine_xy_goal_tolerance` to a reasonable distance and saw from the logs that the goal checker succeeded by using `fine_xy_goal_tolerance` 
- Set the `fine_xy_goal_tolerance` very small and saw from the logs that the goal checker succeeded by using `coarse_xy_goal_tolerance` 

That's the basic behavior I was looking for :+1: I didn't test every edge case though. 

Maybe let's not immediately set this as the default - Once it's merged, I will try to see if we can have it to be the default internally for some time (+ hopefully other people test) and after this we can set it as the default.

 @doisyg FYI, let's discuss if this has any drawbacks for us

**Comment by SteveMacenski** [comment]
@DavidG-Develop OK! Final steps:

* We need to add to the migration guide that this new goal checker exists
* We need to add to the configuration guide for the controller server the new goal checker
* Do some final tests in simulation using MPPI and  RPP (~20 paths achieved at various angles each) to show that we correctly continue to go to the refined until we stall, achieve the fine bounds, and/or pass the finish line.  Use TB3/4 sim from nav2 bringup.  I basically just want you to sanity check the course-stopped, fine-achieved, or finish-line crossed behaviors using the full system to make sure there aren't unintended issues that are easily predictable from non-unit testing. For each situation at the end, make sure the right one triggered and the behavior was good with the counters (maybe we want to retune the default values?) - adding some temporary logging in the plugin may help in making this clear.  

Then we can merge! Thanks for all this great work and would love your help in more tickets if you're open to it!

**Comment by DavidG-Develop** [comment]
@SteveMacenski I will look that I get the docs written sometime next week. 

For the testing, I did already do some testing on the tb4 sim by sending goals with poses tight next to obstacles, and I did observe all 4 cases proc at least once. I do also agree we should check and possibly tune the default values. To do that I would most likely build some evaluation script where I can than track the trajectory of the robot as well as I can visualise the finish line to see if all triggers correctly... Any recommendations for initial default values to start testing from?
> 
> Then we can merge! Thanks for all this great work and would love your help in more tickets if you're open to it!

Thanks for the kind feedback, I really appreciate it! Also thanks for the support on this one 😄 ! I usually would like to take on 1 ticket per month where I can then slowly work on it whenever I catch some time, but due to a full-time job, an extra part-time job at my Uni and always doing stuff for our Robocup team, the rest free time is mostly not enought for a ticket per month 🤣 . (just went back to check my last merge was sep 2025... ooops haha, so once a month was myb slightly extended 🤣 🤣 )

**Comment by elsayedelsheikh** [comment]
Great work! 
Is it worth adding additional check for `fine_xy_goal_tolerance_ > coarse_xy_goal_tolerance_` on `updateParametersCallback()` ?

**Comment by DavidG-Develop** [comment]
@elsayedelsheikh I don't find it necessary... If someone sets `fine_xy_goal_tolerance_ > coarse_xy_goal_tolerance_` than the checker would just act as a simple goal checker (maybe it can also be seen as an additional feature that the user can switch it during runtime from adaptive to simple). I could add a warning print to notify the user that that configuration is not intended, but I do not see a reason why it should be considered "wrong". Thoughts @SteveMacenski ?

**Comment by elsayedelsheikh** [comment]
> @elsayedelsheikh I don't find it necessary... If someone sets `fine_xy_goal_tolerance_ > coarse_xy_goal_tolerance_` than the checker would just act as a simple goal checker (maybe it can also be seen as an additional feature that the user can switch it during runtime from adaptive to simple). I could add a warning print to notify the user that that configuration is not intended, but I do not see a reason why it should be considered "wrong". Thoughts @SteveMacenski ?

A warning would be nice!

**Comment by DavidG-Develop** [comment]
So I added the warn for coarse < fine, but this got me thinking for other params as well... 

So currently I followed the setup of other checkers and do check that all doubles >= 0 and all ints >= 1 but only on the params callback and not on init... Now my question would be is it better that I follow the way the others are structured and do not check all params in init (assuming the user will not put in *unreasonable* values) , or would it better be to actually check them in init as well and either throw or just warn and use default? I made sure the code does not crash even with values which are not intended to be used but it does make it behave not as intended (never succeeding, succeeding instantly, and so on).

Also I noticed we never use result.reason in the `validateParameterUpdatesCallback` which I find to be sub optimal, but again I went with following the other checkers structure but I believe it would be better if we also populate that field with the reason before returning.

Thoughts? @SteveMacenski @elsayedelsheikh

**Comment by elsayedelsheikh** [comment]
In my opinion, users often overlook dynamic parameter updates, so a warning is necessary.
On initialization, they’re more likely to follow the documentation/guides.
Let's hear from @SteveMacenski

**Comment by DavidG-Develop** [comment]
Okay so I built a script which does some testing and the following are the results:
MPPI default settings:
- Not close to obstacles always reaching fine tolerance
- Close to obstacles - 80% fine, 14% stop triggered, 6% distance triggered



To test the finish line I had to force it by setting params in a way no normal person should but also seems to trigger correctly
image

Now with RRP the story is much different... To start off I have never used RRP so my knowledge of the control algorithm behind it is 0! With me using default params even without obstacles it would 100% of the times trigger the distance stagnation due to it stopping short... 
image

Now as I am not expirienced with it I would leave it up to you guys to check/decide if this is fine or it should be changed...

From my testes based on TB4 sim the default values are fine, but ofc for larger and slower robots they will have to be adjusted...

**Comment by SteveMacenski** [comment]
@DavidG-Develop please rebase/pull in main in order to have CI turn over. 

I think everything was answered in your exchanges so I don't think there's anything for me to add more :-) Just the documentation!

**Comment by SteveMacenski** [comment]
@elsayedelsheikh did you give this a nice test / evaluation? I want to merge this in this week, but I would like ~2-3 outside parties to corroborate that it works well for their applications and gets better results (and no edge cases that we didn't handle)

**Comment by elsayedelsheikh** [comment]
> @elsayedelsheikh did you give this a nice test / evaluation? I want to merge this in this week, but I would like ~2-3 outside parties to corroborate that it works well for their applications and gets better results (and no edge cases that we didn't handle)

I've run a few tests in simulation, and so far, so good. I'll continue tomorrow with more testing on a real robot, and if there's any problem, I'll let you know by the end of the day.  :-)

**Comment by SteveMacenski** [comment]
@DavidG-Develop if you get the docs together, I can merge this by Friday!

**Comment by DavidG-Develop** [comment]
@SteveMacenski Bruh you merged the docs already 🤣 🤣 I see why you asked on Linked to possibly find some other ppl who can also do the testing, aproving and merging 😆

**Comment by SteveMacenski** [comment]
:upside_down_face: sorry... I juggle alot of things and sometimes I forget the state of them 

I think just some tests by the community to get a :+1: and I'll merge!

**Comment by elsayedelsheikh** [comment]
I don’t even know how @SteveMacenski keeps track of all these open tickets. :muscle: 
I can barely remember what I had for breakfast :sweat_smile: 
 
P.S. My schedule finally works, and I’ve set aside 2 hours daily for Nav2. Feel free to mention me on any tickets.

**Comment by SteveMacenski** [comment]
How'd the hardware testing go @elsayedelsheikh?

**Comment by elsayedelsheikh** [comment]
@SteveMacenski 
Unfortunately, I didn’t get the chance to test it on hardware since some motor driver issues showed up earlier today.

Anyway, this LGTM! I ran more simulation scenarios with different goal headings and also very close goals where making those turns is basically not possible, while keeping the goal checker fine tolerance quite tight so coarse tolerance shows its powers and It’s working really well with MPPI :100: 

> Maybe let's not immediately set this as the default - Once it's merged, I will try to see if we can have it to be the default internally for some time (+ hopefully other people test) and after this we can set it as the default.

+1

**Comment by SteveMacenski** [comment]
Thanks @DavidG-Develop for your help -- would welcome your assistence on other tickets too (or other things you might want to improve) if you wanted and had the time!

**Comment by DavidG-Develop** [comment]
Hey @SteveMacenski I will try to take on an issue whenever I catch enough free time... In the meantime, you can always ping me for input or a quick sim test!
